# Assignment 4

## Requirements
We are going to use ALL the features stored in x to do regression and we are NOT going to make any plots this time. 

Split your x and y data into training and test sets

Train the four regression models using your training data
Evaluate the MSE scores of the models on your test data

Use cross-validation with your training data to select the optimal values of several hyperparameters for your models

## Set Up

In [53]:
import sklearn.datasets
import numpy as np

diabetes = sklearn.datasets.load_diabetes()
print(diabetes.DESCR)
x = diabetes.data
y = diabetes.target

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

## K Nearest Neighbours

Use n_neighbors for k-nearest neighbors

In [22]:
import sklearn.neighbors
import sklearn.model_selection
from sklearn.metrics import mean_squared_error

modelkn = sklearn.neighbors.KNeighborsRegressor(n_neighbors=10)
x_train, x_test, y_train, y_test = sklearn.model_selection.train_test_split(
    x, y, test_size=0.2, random_state=42)

modelkn.fit(x_train, y_train)
y_predkn = modelkn.predict(x_test)

print('MSE_knn = ', mean_squared_error(y_test, y_predkn))

MSE_knn =  3115.359438202248


## Linear Regression

In [24]:
import sklearn.linear_model

modellr = sklearn.linear_model.LinearRegression()
x_trainlr, x_testlr, y_trainlr, y_testlr = sklearn.model_selection.train_test_split(
    x, y, test_size=0.2, random_state=42)

modellr.fit(x_trainlr, y_trainlr)
y_predlr = modellr.predict(x_testlr)

print('MSE_knn = ', mean_squared_error(y_testlr, y_predlr))

MSE_knn =  2900.1732878832318


## Decision Tree Regression

Use max_depth for decision tree

In [34]:
import numpy as np
import matplotlib.pyplot as plt
import sklearn.tree

from sklearn.metrics import mean_squared_error

model3 = sklearn.tree.DecisionTreeRegressor(max_depth=10)

x_train3, x_test3, y_train3, y_test3 = sklearn.model_selection.train_test_split(
    x, y, test_size=0.2, random_state=42)

model3.fit(x_train3,y_train3)
y_pred3 = model3.predict(x_test3)
print('MSE_tree = ', mean_squared_error(y_test3, y_pred3))

MSE_tree =  4080.591306373232


## Random Forest Regression

Use max_depth for random forest

In [38]:
from sklearn.ensemble import RandomForestRegressor
model4 = sklearn.ensemble.RandomForestRegressor(max_depth=10)

x_train4, x_test4, y_train4, y_test4 = sklearn.model_selection.train_test_split(
    x, y, test_size=0.2, random_state=42)

model4.fit(x_train4, y_train4)
y_pred4 = model4.predict(x_test4)

print('MSE_depth10 = ', mean_squared_error(y_test, y_pred3))

MSE_depth10 =  4080.591306373232


## Best Performing Regression based on MSE 

KNN MSE: 3115

Linear Regression MSE: 2900

Decision Tree MSE: 4080

Random Forest MSE: 4080

The smaller the MSE, the more accurate the data. Thus, Linear Regression has the lowest MSE and thus is the best performing regression.

## Tree Visualization 

#### Use a tree visualization method to show the trained decision tree model

In [55]:
text_representation = tree.export_text(model3)
print(text_representation)

|--- feature_2 <= 0.01
|   |--- feature_8 <= 0.01
|   |   |--- feature_8 <= -0.04
|   |   |   |--- feature_4 <= -0.04
|   |   |   |   |--- feature_4 <= -0.05
|   |   |   |   |   |--- feature_9 <= -0.08
|   |   |   |   |   |   |--- feature_6 <= 0.01
|   |   |   |   |   |   |   |--- value: [70.00]
|   |   |   |   |   |   |--- feature_6 >  0.01
|   |   |   |   |   |   |   |--- feature_2 <= -0.05
|   |   |   |   |   |   |   |   |--- value: [143.00]
|   |   |   |   |   |   |   |--- feature_2 >  -0.05
|   |   |   |   |   |   |   |   |--- feature_3 <= -0.03
|   |   |   |   |   |   |   |   |   |--- value: [114.00]
|   |   |   |   |   |   |   |   |--- feature_3 >  -0.03
|   |   |   |   |   |   |   |   |   |--- value: [125.00]
|   |   |   |   |   |--- feature_9 >  -0.08
|   |   |   |   |   |   |--- feature_0 <= 0.01
|   |   |   |   |   |   |   |--- feature_2 <= -0.08
|   |   |   |   |   |   |   |   |--- feature_6 <= 0.03
|   |   |   |   |   |   |   |   |   |--- feature_4 <= -0.08
|   |   |   |  

In [ ]:
### Use a tree visualization method to show one of the trees of the random forest model

In [56]:
model4.estimators_

[DecisionTreeRegressor(max_depth=10, max_features='auto',
                       random_state=1116308189),
 DecisionTreeRegressor(max_depth=10, max_features='auto',
                       random_state=1603317762),
 DecisionTreeRegressor(max_depth=10, max_features='auto',
                       random_state=1147781571),
 DecisionTreeRegressor(max_depth=10, max_features='auto', random_state=445378186),
 DecisionTreeRegressor(max_depth=10, max_features='auto', random_state=707737843),
 DecisionTreeRegressor(max_depth=10, max_features='auto',
                       random_state=1836321331),
 DecisionTreeRegressor(max_depth=10, max_features='auto',
                       random_state=1846863582),
 DecisionTreeRegressor(max_depth=10, max_features='auto', random_state=54036853),
 DecisionTreeRegressor(max_depth=10, max_features='auto',
                       random_state=1900879465),
 DecisionTreeRegressor(max_depth=10, max_features='auto', random_state=119076696),
 DecisionTreeRegressor(max_